Source/HTTP/Curl/CurlEasyRequest.cpp

```diff
@@ -0,0 +1,345 @@
+    }
+
+    // url & method
+    char const* url = nullptr;
+    char const* method = nullptr;
+    RETURN_IF_FAILED(HCHttpCallRequestGetUrl(hcCall, &method, &url));
+    RETURN_IF_FAILED(easyRequest->SetOpt<char const*>(CURLOPT_URL, url));
+
+    CURLoption opt = CURLOPT_HTTPGET;
+    RETURN_IF_FAILED(MethodStringToOpt(method, opt));
+    if (opt == CURLOPT_CUSTOMREQUEST)
+    {
+        RETURN_IF_FAILED(easyRequest->SetOpt<char const*>(opt, method));
+    }
+    else
+    {
+        RETURN_IF_FAILED(easyRequest->SetOpt<long>(opt, 1));
+    }
+
+    // headers
+    uint32_t headerCount{ 0 };
+    RETURN_IF_FAILED(HCHttpCallRequestGetNumHeaders(hcCall, &headerCount));
+
+    for (auto i = 0u; i < headerCount; ++i)
+    {
+        char const* name{ nullptr };
+        char const* value{ nullptr };
+        RETURN_IF_FAILED(HCHttpCallRequestGetHeaderAtIndex(hcCall, i, &name, &value));
+        RETURN_IF_FAILED(easyRequest->AddHeader(name, value));
+    }
+    RETURN_IF_FAILED(easyRequest->SetOpt<curl_slist*>(CURLOPT_HTTPHEADER, easyRequest->m_headers));
+
+    // timeout
+    uint32_t timeoutSeconds{ 0 };
+    RETURN_IF_FAILED(HCHttpCallRequestGetTimeout(hcCall, &timeoutSeconds));
+    RETURN_IF_FAILED(easyRequest->SetOpt<long>(CURLOPT_TIMEOUT_MS, timeoutSeconds * 1000));
+
+    RETURN_IF_FAILED(easyRequest->SetOpt<long>(CURLOPT_VERBOSE, 0)); // verbose logging (0 off, 1 on)
+    RETURN_IF_FAILED(easyRequest->SetOpt<long>(CURLOPT_HEADER, 0)); // do not write headers to the write callback
+    RETURN_IF_FAILED(easyRequest->SetOpt<char*>(CURLOPT_ERRORBUFFER, easyRequest->m_errorBuffer));
+
+    // write data callback
+    RETURN_IF_FAILED(easyRequest->SetOpt<curl_write_callback>(CURLOPT_WRITEFUNCTION, &WriteDataCallback));
+    RETURN_IF_FAILED(easyRequest->SetOpt<void*>(CURLOPT_WRITEDATA, easyRequest.get()));
+
+    // write header callback
+    RETURN_IF_FAILED(easyRequest->SetOpt<curl_write_callback>(CURLOPT_HEADERFUNCTION, &WriteHeaderCallback));
+    RETURN_IF_FAILED(easyRequest->SetOpt<void*>(CURLOPT_HEADERDATA, easyRequest.get()));
+
+    // debug callback
+    RETURN_IF_FAILED(easyRequest->SetOpt<curl_debug_callback>(CURLOPT_DEBUGFUNCTION, &DebugCallback));
+
+    return Result<HC_UNIQUE_PTR<CurlEasyRequest>>{ std::move(easyRequest) };
+}
+
+CURL* CurlEasyRequest::Handle() const noexcept
+{
+    return m_curlEasyHandle;
+}
+
+void CurlEasyRequest::Complete(CURLcode result)
+{
+    HC_TRACE_VERBOSE(HTTPCLIENT, "CurlEasyRequest::Complete: CURLCode=%ul", result);
+
+    long httpStatus = 0;
+    auto getInfoRes = curl_easy_getinfo(m_curlEasyHandle, CURLINFO_RESPONSE_CODE, &httpStatus);
+    if (getInfoRes != CURLE_OK)
+    {
+        result = getInfoRes;
+    }
+
+    if (result != CURLE_OK)
+    {
+        HC_TRACE_VERBOSE(HTTPCLIENT, "CurlEasyRequest::m_errorBuffer='%s'", m_errorBuffer);
+
+        HRESULT hr = HCHttpCallResponseSetNetworkErrorCode(m_hcCallHandle, E_FAIL, result);
```

Comment:
You can get the underlying WinHTTP HRESULT error with CURLINFO_OS_ERRNO.  It's super useful for debugging issues at least from our end.

Reply:
Also, the one "special suspend" thing with XCurl is that there is a special error for "failed to due to suspend" that should be documented out of libhttpclient/xsapi/etc.